prog/kubernetes/provision_kubernetes_node.rb

```diff
@@ -0,0 +1,134 @@
+  subject_is :kubernetes_cluster
+
+  def vm
+    @vm ||= Vm[frame.fetch("vm_id", nil)] || nil
+  end
+
+  def write_hosts_file_if_needed(ip = nil)
+    return unless Config.development?
+    return if vm.sshable.cmd("sudo cat /etc/hosts").match?(/#{kubernetes_cluster.endpoint}/)
+    ip = kubernetes_cluster.cp_vms.first.ephemeral_net4 if ip.nil?
+
+    vm.sshable.cmd("echo '#{ip} #{kubernetes_cluster.endpoint}' | sudo tee -a /etc/hosts")
+  end
+
+  label def start
+    name = "#{kubernetes_cluster.name.downcase}-control-plane-#{SecureRandom.alphanumeric(5).downcase}"
+
+    vm_st = Prog::Vm::Nexus.assemble_with_sshable(
+      "ubi",
+      kubernetes_cluster.project.id,
+      # we should reiterate how we name the vm. some how correlate it to the vm's info.
+      name: name,
+      location: kubernetes_cluster.location,
+      size: "standard-2",
+      boot_image: "ubuntu-jammy",
+      private_subnet_id: kubernetes_cluster.private_subnet_id,
+      enable_ip4: true
+    )
+
+    current_frame = strand.stack.first
+    current_frame["vm_id"] = vm_st.id
+    strand.modified!(:stack)
+
+    kubernetes_cluster.add_cp_vm(vm_st.subject)
+    kubernetes_cluster.api_server_lb.add_vm(vm_st.subject)
+
+    hop_install_software
+  end
+
+  label def install_software
+    nap 5 unless vm.strand.label == "wait"
+    vm.sshable.cmd <<BASH
+  set -ueo pipefail
+  echo "net.ipv6.conf.all.forwarding=1\nnet.ipv6.conf.all.proxy_ndp=1\nnet.ipv4.conf.all.forwarding=1\nnet.ipv4.ip_forward=1" | sudo tee /etc/sysctl.d/72-clover-forward-packets.conf
+  sudo sysctl --system
+  sudo apt update
+  sudo apt install -y ca-certificates curl apt-transport-https gpg
+  sudo install -m 0755 -d /etc/apt/keyrings
+  sudo curl -fsSL https://download.docker.com/linux/ubuntu/gpg -o /etc/apt/keyrings/docker.asc
+  sudo chmod a+r /etc/apt/keyrings/docker.asc
+  echo "deb [arch=$(dpkg --print-architecture) signed-by=/etc/apt/keyrings/docker.asc] https://download.docker.com/linux/ubuntu \\$(. /etc/os-release && echo "\\$VERSION_CODENAME") stable" \\| sudo tee /etc/apt/sources.list.d/docker.list > /dev/null
+  sudo apt update
+  sudo apt install -y containerd
+  sudo mkdir -p /etc/containerd
+  sudo touch /etc/containerd/config.toml
+  containerd config default | sudo tee /etc/containerd/config.toml
+  sudo sed -i 's/SystemdCgroup = false/SystemdCgroup = true/' /etc/containerd/config.toml
+  sudo systemctl restart containerd
+  curl -fsSL https://pkgs.k8s.io/core:/stable:/#{kubernetes_cluster.kubernetes_version}/deb/Release.key | sudo gpg --dearmor -o /etc/apt/keyrings/kubernetes-apt-keyring.gpg
+  echo 'deb [signed-by=/etc/apt/keyrings/kubernetes-apt-keyring.gpg] https://pkgs.k8s.io/core:/stable:/#{kubernetes_cluster.kubernetes_version}/deb/ /' | sudo tee /etc/apt/sources.list.d/kubernetes.list
+  sudo apt update
+  sudo apt install -y kubelet kubeadm kubectl
+  sudo apt-mark hold kubelet kubeadm kubectl
+  sudo systemctl enable --now kubelet
```

Comment:
I feel like it is better to move this into dataplane and use daemonizer to setup. Also, please make sure this is an idempotent state. A simple network connection issue may leave the strand infinitely running. Considering you `push` this prog, clean-up is a very hairy manual labor.

Reply:
This is a temporary step that makes provisioning less resilient and more time-consuming. We are going to create VM images for k8s nodes but IMO this is OK for now to save time to start getting some feedback.